### DISCUSSION TITLE
How to keep a user signed in after auth

### DISCUSSION THREAD

**Discussion by kaboc** [pocketbase/pocketbase#2154] [Q&A] [answered]
I created a custom `AuthStore` and checked the behaviour of `authRefresh()` in Dart. It turned out that it worked successfully while the token was still valid, but threw an exception if the token had already expired. Is it not possible to maintain a signed-in state forever?

I think it is quite common that users can keep using a mobile app without signing in again even after not using the app for a very long time, and it is possible with Flutter + email auth of Firebase and Supabase. I'd appreciate some advice.

**Comment by ganigeorgiev** [comment, marked as answer]
Could you share your custom AuthStore? What do you use for persistence - encrypted shared_preferences?

**Reply by kaboc** [reply]
Here you are.

```dart
import 'dart:convert';

import 'package:hive/hive.dart';
import 'package:pocketbase/pocketbase.dart';

class CustomAuthStore extends AuthStore {
  CustomAuthStore._();

  static CustomAuthStore? _instance;
  late final Box<String> _box;

  static Future<CustomAuthStore> initialize() async {
    if (_instance == null) {
      Hive.init('path/to/box/dir');
      _instance = CustomAuthStore._().._box = await Hive.openBox('session');
    }
    final instance = _instance!;

    instance.onChange.listen((e) async {
      final model = e.model;
      if (model == null || !instance.isValid) {
        await instance._box.clear();
      } else if (model is RecordModel) {
        await instance._box.putAll({
          'token': e.token,
          'model': jsonEncode(model),
        });
      }
    });

    return instance;
  }

  Future<bool> restore({required PocketBase pb}) async {
    final token = _box.get('token');
    final model = _box.get('model');

    if (token == null) {
      return false;
    }

    save(
      token,
      model == null ? null : RecordModel.fromJson(jsonDecode(model)),
    );

    try {
      await pb.collection('users').authRefresh();
      return isValid;
    } on Exception {
      return false;
    }
  }
}
```

```dart
final authStore = await CustomAuthStore.initialize();
final pb = PocketBase(
  'http://127.0.0.1:8090',
  authStore: authStore,
);

final success = await authStore.restore(pb: pb);
```

I'm using Hive without encryption as an experiment to see how things work. I'll probably enable encryption and may use a different storage.

**Reply by ganigeorgiev** [reply]
Hm, the above looks fine as long as the Hive data are actually persisted on the disk.

Is there any specific error that the above is throwing?

PocketBase auth tokens are usually long lived (~14days by default). When calling `authRefresh()` the store gets updated with a token with new `exp` claim. If you want to increase the default token duration, you can change it from:
_"Admin UI > Settings > Token options > Auth record authentication token"_


---

_Side-note: One minor optimization you can do is to eagerly check `isValid` before making the `authRefresh` call to avoid an unnecessary request:_

```dart
if (!isValid) {
  return false
}
await pb.collection('users').authRefresh();
```

**Reply by kaboc** [reply]
I've changed the setting to make a token expire in two minutes so that I can check the behaviour again and again. My custom AuthStore is working fine and a token is refreshed successfully as long as it is done before expiration. My issue is that users are signed out if they don't use the app for a longer period of time than the token duration.

Is there no option but to make the duration extremely long? Using a long lived token does not seem safe to me. It would be great if the authentication with email and password worked like OAuth 2.0, allowing a token (an access token) to be refreshed with a refresh token regardless of whether the access token expired or not. In that case, an access token can be made short lived and becomes safer.

> Side-note: One minor optimization you can do is to eagerly check isValid before making the authRefresh call to avoid an unnecessary request

Thank you! I'll add the check.

**Reply by ganigeorgiev** [reply]
>  Using a long lived token does not seem safe to me. It would be great if the authentication with email and password worked like OAuth 2.0, allowing a token (an access token) to be refreshed with a refresh token regardless of whether the access token expired or not. In that case, an access token can be made short lived and becomes safer.

How a long lived refresh token is more safer than a long lived auth token if both are stored on the same place? What if an attacker get access to your refresh token? For OAuth2 it make sense to have both tokens, because usually they are associated to certain scopes, but PocketBase doesn't have such concept. You are either authenticated or not. The access control is based on the individual collection API rules.
Another reason why you may want to have 2 tokens could be also to limit the number of services to which you send the tokens (eg. the refresh token can be send only to the auth server, while the short lived access token can be send to one or multiple app servers), but again, this is not the case for PocketBase because all tokens will end up in 1 place.

Also please note that the PocketBase API is completely stateless, meaning that the tokens are not stored in the database.

> Is there no option but to make the duration extremely long? 

I can't think of other option. Storing the token data in an ["Encrypted box"](https://docs.hivedb.dev/#/advanced/encrypted_box) and having a duration for a week or two should be enough in my opinion.

**Reply by kaboc** [reply]
What wrote about safety is based on my shallow knowledge. I've learned in the past that the combination of a short lived auth token and a long lived refresh token if for better safety. I didn't think it would make you so unhappy. I apologise.

I now understand the email auth of PocketBase has a very different mechanism from OAuth 2 and it is not possible to prevent users from signing out automatically.

I have some apps I don't use for more than two weeks (or even much longer) on my phone, but I'm still signed in next time, except for a few apps like one for banking, which needs very strict security. Therefore I think it is unusual that a user is signed out so soon. If the token duration shouldn't be longer than two weeks or so in PocketBase, I'll think about avoiding the email auth and using only OAuth providers.

**Reply by ganigeorgiev** [reply]
>  I didn't think it would make you so unhappy. I apologise.

Sorry if my reply came of rude, I'm not a native speaker and it definitely wasn't intentional.

I just don't understand why do you think a refresh token would be more secure, considering how PocketBase manage tokens.

You can treat the long lived auth token the same way as the the long lived refresh token.

**Reply by kaboc** [reply]
I don't clearly remember the reason I read somewhere before, but it was probably like, a refresh token is not included in each communication and it is only used for refreshing the access token, so there is little danger that it is stolen during API access. I don't know whether it was from a reliable source, but it made me think it would be unsafe to use an auth token in the same way as a refresh token. It may be wrong.

**Reply by ganigeorgiev** [reply]
This is similar to the second case outlined in https://github.com/pocketbase/pocketbase/discussions/2154#discussioncomment-5425468.

It makes sense for example if you have a dedicated auth microservice and several other applications that shares the same auth data to use the refresh token only with the auth service and the access token for granting access to the other applications, but that's not the case with PocketBase because it is a monolith and everything happens in one place on your server. Additionally, as mentioned earlier, **PocketBase tokens are not stored on the server**, so even if you generate 1k tokens for 1 user, they will exist only on the user device and during the encrypted (if you use https) network request-response cycle.

Or in other words, if you have MITM and if someone can steal your access token, they can also just as easy steal your refresh token when you send the refresh call every time when the user open the app. For PocketBase a long lived auth token is no more or less secure than a long lived refresh token.

**Reply by kaboc** [reply]
Yes, actually I had already understood it from your previous messages. Thank you for making it even clearer.

So it is definitely not appropriate to request you to support what I was wanting. I'll think about how I can manage to make my future apps feasible within the range of features provided by PocketBase.

Thank you for all the advices you gave me!

**Reply by lukepighetti** [reply]
@ganigeorgiev whats the best way to make sure someone doesn't have to sign back in more than once a year, while making sure there aren't old valid tokens floating around that are good for a year?

Using firebase_auth I've noticed that logged in accounts can stay active for multiple years and it's a really nice user experience. Being logged out even once a month is not a great experience.

Also curious how we can invalidate tokens remotely. Consider the ability to see authenticated devices and removing ones you don't recognize.

That all said, I agree that simplicity is a feature.

**Reply by ganigeorgiev** [reply]
@lukepighetti I'm failing to understand the question. In general the default tokens duration can be changed from the _Admin UI > Settings > Token options_.

> while making sure there aren't old valid tokens floating around that are good for a year?

There is no such option.

> Using firebase_auth I've noticed that logged in accounts can stay active for multiple years and it's a really nice user experience. Being logged out even once a month is not a great experience.

This is probably because Firebase refresh tokens practically never expires on their own. From the [Firebase docs](https://firebase.google.com/docs/auth/admin/manage-sessions):

> Refresh tokens expire only when one of the following occurs:
> - The user is deleted 
> - The user is disabled
> - A major account change is detected for the user. This includes events like password or email address updates.

Once again, PocketBase doesn't have the concept of access+refresh token because it doesn't need it for the reasons explained above. A long lived Firebase refresh token serves the same purpose as long-lived PocketBase auth token.

> Also curious how we can invalidate tokens remotely. 

As mentioned above, PocketBase is fully stateless and tokens are not stored anywhere, meaning that individual token can't be invalidated. You can invalidate only:

- all issued tokens for all users (from the _Admin UI > Settings > Token options_)
- all issued tokens for a single user (this happens automatically on confirmed password or email change; we don't have exposed Web API to trigger the individual user invalidation manually at the moment - to some extend this could be emulated by showing a "Confirm invalidation" screen to the user and behind the scene you send a password change update with the provided password/in this case `oldPassword`, `password` and `passwordConfirm` will have the same value/).

**Reply by ganigeorgiev** [reply]
Also note that the latest Dart SDK comes with a helper `AsyncAuthStore` that should help in reducing the required boilerplate to integrate with various async storage mechanisms. You could find an example in https://github.com/pocketbase/dart-sdk?tab=readme-ov-file#authstore.

**Comment by kaboc** [comment]
I removed "with email and password" from the title.

I had been thinking that a user would need to be reauthenticated if the user session had already expired, but I was wrong in thinking that it was only about the auth with an email. I recently implemented authentication with providers too and realised that the same applied to those actually.

My understanding is that providers are only used for signing up/in a user to PocketBase. A token created by PocketBase (not by the provider) is used from that point onwards. It has all necessary info like the the expiry date in its payload because authentications in PocketBase is stateless, and there are no such things as an access token and a refresh token. So a user session is only active while the token is valid, and thus a user needs to refresh the session before expiration. Refreshing is a process only within PocketBase, without providers involved.

I think understanding it is very important for developers, but in my opinion, the "Authentication" page in the document hasn't explained it enough. That is why I thought I should leave this note here for others users. It'd be great if the documentation was improved.
